packages/tpetra/core/example/Finite-Element-Assembly/fem_assembly_type3_staticprofile.cpp

```diff
@@ -0,0 +1,322 @@
+// @HEADER
+// ***********************************************************************
+//
+//          Tpetra: Templated Linear Algebra Services Package
+//                 Copyright (2008) Sandia Corporation
+//
+// Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ************************************************************************
+// @HEADER
+#include <cmath>
+#include <iostream>
+#include <iomanip>
+#include <sstream>
+
+#include <Tpetra_DefaultPlatform.hpp>
+#include <Tpetra_Version.hpp>
+#include <MatrixMarket_Tpetra.hpp>
+#include <Teuchos_RCP.hpp>
+#include <Teuchos_GlobalMPISession.hpp>
+#include <Teuchos_FancyOStream.hpp>
+
+#include "typedefs.hpp"
+#include "MeshDatabase.hpp"
+#include "Element.hpp"
+
+#define PRINT_VERBOSE 0
+
+
+
+int main (int argc, char *argv[]) 
+{
+  using Teuchos::RCP;
+  using Teuchos::TimeMonitor;
+
+  const GlobalOrdinal GO_INVALID = Teuchos::OrdinalTraits<GlobalOrdinal>::invalid();
```

Comment:
I'm not a fan of having the typedef name (`GlobalOrdinal`) be the same as the template parameter name.  Try `GO` or `global_ordinal_type` or something else.

Reply:
I know MueLu does this but I find that super confusing.

Reply:
👍  Changed all the typedefs in typedefs.hpp in my new branch to use a `typename_t` scheme and be consistent.  There was actually a mix of lowercase `typename_type` and camelcase `FooType` style entries.

Reply:
lower-case `typename_type` is how Tpetra classes do their typedefs.  CamelCase FooType is how Tpetra classes do their template parameters.

Reply:
`GO` etc. are acceptable abbreviations for `global_ordinal_type` etc.

Reply:
Does a style guide exist anywhere?  One thing I've bumped against for Trilinos so far is that there doesn't appear to be any style guide that has any official weight... mostly it appear that folks write loosely conforming code but aesthetics can vary quite a bit between developers.

Anyways... if a style suggestions/guide lives somewhere, I'd love a link to it :D

I've always liked either having `_type`, `_t`, or `CamelCaseType` as the suffix on types that way I know exactly that it's a typename. 

I'm actually not a fan, personally, of having typenames in all caps because I've always seen that as some kind of constant (either const, or preprocessor).

I'm happy to make these things conform to whatever the standard is... but I don't want to change all 10(ish) of the declarations in `typedefs.hpp` again b/c it's fairly tedious ;)

Reply:
@william76 Don't worry about it then.